Documentation: Add details on jsonpath filter (#858)
RedisJSON/RedisJSON: docs/docs/path.md

"[\"Noise-cancelling Bluetooth headphones\",\"Wireless earbuds\"]"
```

Filter expressions `?()` let you select JSON elements based on certain conditions. You can use comparison operators (`==`, `!=`, `<`, `<=`, `>`, `>=`, `=~`), logical operators (`&&`, `||`), and parenthesis (`(`, `)`) within these expressions. A filter expression can be applied on an array or on an object, iterating all the **elements** in the array or all the **values** in the object, retrieving only the ones that match the filter condition. 

Paths within the filter condition are using the dot notation with either `@` to denote the current array element or the current object value, or `$` to denote the top-level element. For example, `@.key_name` to refer to a nested value and `$.top_level_key_name` to refer to a top-level value.

> The comparison operator `=~` is matching a path of a string value on the left-hand side against a regular expression pattern on the right-hand side. The supported regular expression syntax is detailed [here](https://docs.rs/regex/latest/regex/#syntax).

Reviewer: ```suggestion
The comparison operator `=~` is matching a path of a string value on the left side against a regular expression pattern on the right side.  For more information, see the [supported regular expression syntax docs](https://docs.rs/regex/latest/regex/#syntax).
```

> Non-string values do not match. A match can only occur when the left-hand side is a path of a string value and the right-hand side is either a hard-coded string, or a path of a string value. See [examples](#json-filter-examples) below.

Reviewer: ```suggestion
Non-string values do not match. A match can only occur when the left side is a path of a string value and the right side is either a hard-coded string, or a path of a string value. See [examples](#json-filter-examples) below.
```

The regex match is partial, meaning `"foo"` regex pattern matches a string such as `"barefoots"`.
To make it exact, use the regex pattern `"^foo$"`.

Other JSONPath engines may use regex pattern between slashes, e.g., `/foo/`, and their match is exact.

> They can perform partial match using a regex pattern such as `/.*foo.*/`.

Reviewer: ```suggestion
They can perform partial matches using a regex pattern such as `/.*foo.*/`.
```

#### JSON Filter examples

> In the following example, this filter only returns wireless headphones with a price less than 70:

Reviewer: ```suggestion
In the following example, the filter only returns wireless headphones with a price less than 70:
```

"[\"Noise-cancelling Bluetooth headphones\",\"Wireless earbuds\",\"Wireless keyboard\"]"
```

This example filters only keyboards with some sort of USB connection using regex match. Notice this match is case-insensitive thanks to the prefix `(?i)` in the regular expression pattern `"(?i)usb"`:

```sh
127.0.0.1:6379> JSON.GET store '$.inventory.keyboards[?(@.connection =~ "(?i)usb")]'
"[{\"id\":22346,\"name\":\"USB-C keyboard\",\"description\":\"Wired USB-C keyboard\",\"wireless\":false,\"connection\":\"USB-C\",\"price\":29.99,\"stock\":30,\"free-shipping\":false}]"
```

> The regular expression pattern can also be specified using a path of a string value on the right-hand side.

Reviewer: ```suggestion
The regular expression pattern can also be specified using a path of a string value on the right side.
```

OK
```

> Now we can match against the value of `regex_pat` instead of a hard-coded regular expression pattern, and get the keyboard with the `Bluetooth` string in its `connection` key. Notice the one with `USB-C` does not match since its regular expression pattern is case-sensitive and the regular expression pattern is using lower case:

Reviewer: ```suggestion
Now we can match against the value of `regex_pat` instead of a hard-coded regular expression pattern, and get the keyboard with the `Bluetooth` string in its `connection` key. Notice the one with `USB-C` does not match since its regular expression pattern is case-sensitive and the regular expression pattern is using lowercase:
```
